File: plugin/src/main/java/com/xuebinduan/plugin/MainActivity.java
package com.xuebinduan.plugin;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.widget.TextView;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

import dalvik.system.DexClassLoader;
import okio.BufferedSink;
import okio.Okio;
import okio.Source;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        File apk = new File(getCacheDir()+"/plugin.apk");
        if (!apk.exists()){
            try (Source source = Okio.source(getAssets().open("plugin/plugin.apk"));
                 BufferedSink sink = Okio.buffer(Okio.sink(apk))){
                sink.writeAll(source);
            } catch (IOException e) {
                e.printStackTrace();
            }
        }

        if (apk.exists()) {
            try {
                DexClassLoader classLoader = new DexClassLoader(apk.getPath(),getCacheDir().getPath(),null,null);
                Class utilsClass = classLoader.loadClass("com.xuebinduan.myalltest.Utils");
                Constructor utilsConstructor = utilsClass.getDeclaredConstructors()[0];
                utilsConstructor.setAccessible(true);
                Object utils = utilsConstructor.newInstance();
                Method sayMethod = utilsClass.getDeclaredMethod("say");
                //为了调用私有方法
                sayMethod.setAccessible(true);
                String str = (String) sayMethod.invoke(utils);

                ((TextView) findViewById(R.id.text_view)).setText(str);
            } catch (ClassNotFoundException e) {
                e.printStackTrace();
            } catch (IllegalAccessException e) {
                e.printStackTrace();
            } catch (InstantiationException e) {
                e.printStackTrace();
            } catch (InvocationTargetException e) {
                e.printStackTrace();
            } catch (NoSuchMethodException e) {
                e.printStackTrace();
            }
        }
    }

}

File: third_part/moshi/src/main/java/com/xuebinduan/moshi/MainActivity.kt
package com.xuebinduan.moshi

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import com.squareup.moshi.Moshi
import com.squareup.moshi.kotlin.reflect.KotlinJsonAdapterFactory
import com.xuebinduan.moshi.data.Node
import com.xuebinduan.moshi.data.Root
import com.xuebinduan.moshi.data.node.AudioNode


class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        //java object to json
        var root = Root()
        root.data.add(Node())
        root.data.add(AudioNode())

        val moshi: Moshi = Moshi.Builder()
            .addLast(KotlinJsonAdapterFactory())
            .build()

        val jsonAdapter = moshi.adapter(Root::class.java)
        val json = jsonAdapter.toJson(root)
        println(json)


    }
}